server/api/config.go

```diff
@@ -50,31 +52,62 @@ func (h *confHandler) Post(w http.ResponseWriter, r *http.Request) {
 		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
 		return
 	}
-	if err := json.Unmarshal(data, &config.Schedule); err != nil {
+	up1, err := h.updateSchedule(data, config)
+	if err != nil {
 		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
 		return
 	}
-	if err := json.Unmarshal(data, &config.Replication); err != nil {
+	up2, err := h.updateReplication(data, config)
+	if err != nil {
 		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
 		return
 	}
-	if err := json.Unmarshal(data, &config.PDServerCfg); err != nil {
+	up3, err := h.updatePDServerConfig(data, config)
+	if err != nil {
 		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
 		return
 	}
-	if err := h.svr.SetScheduleConfig(config.Schedule); err != nil {
-		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+	if !up1 && !up2 && !up3 {
+		h.rd.JSON(w, http.StatusBadRequest, "config item not found")
 		return
 	}
-	if err := h.svr.SetReplicationConfig(config.Replication); err != nil {
-		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
-		return
+	h.rd.JSON(w, http.StatusOK, nil)
+}
+
+func (h *confHandler) updateSchedule(data []byte, config *config.Config) (bool, error) {
+	old, _ := json.Marshal(config.Schedule)
+	if err := json.Unmarshal(data, &config.Schedule); err != nil {
+		return false, err
 	}
-	if err := h.svr.SetPDServerConfig(config.PDServerCfg); err != nil {
-		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
-		return
+	new, _ := json.Marshal(config.Schedule)
+	if bytes.Equal(old, new) {
+		return false, nil
 	}
-	h.rd.JSON(w, http.StatusOK, nil)
+	return true, h.svr.SetScheduleConfig(config.Schedule)
+}
+
+func (h *confHandler) updateReplication(data []byte, config *config.Config) (bool, error) {
+	old, _ := json.Marshal(config.Replication)
+	if err := json.Unmarshal(data, &config.Replication); err != nil {
+		return false, err
+	}
+	new, _ := json.Marshal(config.Replication)
+	if bytes.Equal(old, new) {
+		return false, nil
+	}
+	return true, h.svr.SetReplicationConfig(config.Replication)
+}
+
+func (h *confHandler) updatePDServerConfig(data []byte, config *config.Config) (bool, error) {
+	old, _ := json.Marshal(config.PDServerCfg)
+	if err := json.Unmarshal(data, &config.PDServerCfg); err != nil {
+		return false, err
+	}
+	new, _ := json.Marshal(config.PDServerCfg)
+	if bytes.Equal(old, new) {
+		return false, nil
+	}
+	return true, h.svr.SetPDServerConfig(config.PDServerCfg)
```

Comment:
It seemed to be not elegant to repeat three times?

Reply:
Yes, I agree with you. But I didn't come up with a good idea to take care of it. 😞